enhancements/workload-partitioning/rationalizing-configuration.md

```diff
@@ -0,0 +1,478 @@
+lot of the work already done in the PAO.
+
+Workload partitioning is currently enabled as part of installing a
+cluster, and that will not change as part of the implementation of
+this design. Enabling partitioning early in this way provides
+predictable behavior from the beginning of the life-cycle, and avoids
+extra reboots required to enable it in a running cluster (an important
+consideration in environments with short, fixed maintenance
+windows). To continue to support enabling the feature during
+deployment, and to simplify the configuration, we propose to extend
+the PAO with a `render` command, similar to the one in other
+operators, to have it generate manifests for the OpenShift installer
+to consume. This avoids any need to change the OpenShift installer to
+support this uncommon use case.
+
+The `render` command will have an option to enable the management
+partition. When the option is given, the `render` command will create
+a manifest for the new enablement API with status data that includes
+the `management` workload type to enable partitioning (the details of
+that API will be defined in a separate design document). In the
+future, the PAO might enable partitioning for all of the types
+mentioned in the input PerformanceProfiles, or based on some other
+input.
+
+For each PerformanceProfile, the PAO will render MachineConfig
+manifests with higher precedence than the default ones created by the
+installer during bootstrapping. These manifests will include all of
+the details for configuring `kubelet`, CRI-O, and `tuned` to know
+about and partition workloads, including the cpusets for each workload
+type.
+
+The PAO `render` command also generates a manifest with the CRD for
+PerformanceProfile so bootstrapping does not block when trying to
+install the PerformanceProfile manifests.
+
+### Future Work
+
+1. Write an enhancement to describe the API the admission hook will
+   use to determine when workload partitioning is enabled.
+2. Write an enhancement to describe the changes in
+   performance-addon-operator to manage the kubelet and CRI-O
+   configuration when workload partitioning is enabled, including the
+   `render` command. See https://issues.redhat.com/browse/CNF-2164 to
+   track that work.
+
+### User Stories
+
+#### Enabling and Configuring at the Same Time
+
+In this workflow, the user provides all of the information needed to
+fully partition the workloads from the very beginning.
+
+1. User runs the installer to `create manifests` to get standard
+   manifests without any partitioning.
+2. The user creates PerformanceProfile manifests for each known
+   MachineConfigPool and adds them to the installer inputs.
+3. The user runs the PAO `render` command to read the
+   PerformanceProfile manifests and create additional manifests, as
+   described above.
+4. User runs the installer to `create cluster`.
+5. Installer bootstraps the cluster.
+6. Bootstrapping runs the machine-config-operator (MCO) `render`
```

Comment:
I'm a bit confused as to what points 6-8 are pointing at. It sounds to me that you are trying to describe the mechanism for which the MCO bootstraps and ensures that the PAO-generated special kubelet/crio configs are ordered. But these steps don't really explicitly highlight that, and also is a bit vague as to the details.

Reply:
Sure :) sorry, I can try to highlight the ordering:

The bootstrap node runs bootstrap versions of 1 MCController 1 MCServer. The MCC is responsible for generating master configs, and the MCS is responsible for serving that master config. 

At the time the MCC runs, it should see all manifests applied to the cluster (as templates directly), at least if it directly targets the MCO namespace. However there is no guarantee that other objects are ready and updated, which is where race conditions have occured in the past: if any of them get modified after bootstrap MCC has finished but the in-cluster MCC has not yet started. (also tangentially related to the kubeletconfig not being day-1 since there is nothing processing those objects into Machineconfigs yet).

The master nodes come up and form the real cluster, at which point it runs the "in-cluster" MCC and MCS, which generates the config to serve worker nodes.

So in that flow, any extra manifests, provided the are created before bootstrap are run, are no different than any template-shipped MCs that are applied day1. The only thing that matters then is the alphanumeric naming schema of the MC object, which determines their order of merging into the "rendered config". That is to say, if 2 MCs named `99-xx` and `11-xx` write to the same file, only the contents of `99-xx` will be there, and whatever is in `11-xx` is "silently dropped"

Reply:
@mrunalp @browsell @MarSik It sounds like we may have to deal with a race condition here.

Reply:
After speaking with @yuqi-zhang offline, I added a note in the risk section to cover this. See fe703ad7924eccf5eb27fcc50bc2ba44235a7727